### PR TITLE
Add parameters to support leader election for Addons Controller

### LINKED ISSUE #2541: Support Leader Election for Addons Controller
(This is used to request new product features)

**Describe the feature request**
In an HA management cluster with addons controller, leader election should be enabled to allow for rapid failover scenarios when some management cluster nodes are down and/or not responding.

**Describe alternatives you've considered**
N/A

**Affected product area (please put an X in all that apply)**

- [ ] APIs
- [ ] Addons
- [ ] CLI
- [ ] Docs
- [ ] IAM
- [ ] Installation
- [ ] Plugin
- [ ] Security
- [ ] Test and Release
- [X] User Experience

**Additional context**
For addons controller deployment onto vSphere, we discovered issues when trying to scale out the addons controller to 3 replicas, and quickly discovered that it wasn't designed to be deployed in an HA configuration. This could potentially be problematic especially if the node the addons-controller pod is scheduled to isn't responding.

### PR DESCRIPTION
**Pull request by haoranleo**
### What this PR does / why we need it
- Add parameters in the addons controller manager's config to support leader election
- Keep leader election option to be false by default (both in manager's config and [addons-manager package](https://github.com/vmware-tanzu/tanzu-framework/blob/main/packages/addons-manager/bundle/config/upstream/addons-manager.yaml#L341))
- Set replicas of deployment `tanzu-addons-controller-manager` to 2 when leader election is enabled

**note:** User can enable leader election by passing in data values when installing package.
### Which issue(s) this PR fixes

Fixes #2541

### Describe testing done for PR
- Passed all tests locally
- Verified deployment `tanzu-addons-controller-manager` yaml is generated correctly based on `leaderElection` setting:
```
# when leaderElection is set to true
spec:
  replicas: 2
  selector:
    matchLabels:
      app: tanzu-addons-manager
  template:
    metadata:
      labels:
        app: tanzu-addons-manager
    spec:
      containers:
      - args:
        - --metrics-bind-addr=localhost:18317
        - --leader-elect=True
        - --health-addr=:18316
        - --webhook-server-port=9865
        - --addon-namespace=tkg-system

# when leaderElection is set to false
spec:
  replicas: 1
  selector:
    matchLabels:
      app: tanzu-addons-manager
  template:
    metadata:
      labels:
        app: tanzu-addons-manager
    spec:
      containers:
      - args:
        - --metrics-bind-addr=localhost:18317
        - --leader-elect=False
        - --health-addr=:18316
        - --webhook-server-port=9865
        - --addon-namespace=tkg-system
```
- Verified the leader election of `addons-controller-manager` works correctly after enabling leader election:
```
# the ConfigMap CR used for holding the leader lock shows up 
[ ~ ]# k get configmap -n vmware-system-tkg | grep tanzu
...
tanzu-addons-manager-leader-lock               0      21s
...

# scale the addons-controller to 3 replicas
[ ~ ]# k get pod -n vmware-system-tkg -o wide | grep addons
tanzu-addons-controller-manager-556c8dddf6-cq89s         1/1     Running   0                61s     192.168.128.1   423bb2f52f8c663136f412787d2345a5   <none>           <none>
[ ~ ]# k scale deploy tanzu-addons-controller-manager -n vmware-system-tkg --replicas=3
deployment.apps/tanzu-addons-controller-manager scaled
[ ~ ]# k get pod -n vmware-system-tkg | grep add
tanzu-addons-controller-manager-556c8dddf6-72m82         1/1     Running   0                24s
tanzu-addons-controller-manager-556c8dddf6-984br         1/1     Running   0                24s
tanzu-addons-controller-manager-556c8dddf6-cq89s         1/1     Running   0                2m57s
```

Related logs of this three replicas:
```
# leader pod
[ ~ ]# k logs tanzu-addons-controller-manager-556c8dddf6-cq89s -n vmware-system-tkg --tail 10
...
I0629 18:28:56.163505       1 leaderelection.go:248] attempting to acquire leader lease vmware-system-tkg/tanzu-addons-manager-leader-lock...
...
I0629 18:28:56.419425       1 leaderelection.go:258] successfully acquired lease vmware-system-tkg/tanzu-addons-manager-leader-lock
...

# non-leader pod1
[ ~ ]# k logs tanzu-addons-controller-manager-556c8dddf6-72m82 -n vmware-system-tkg --tail 10
...
I0629 18:31:27.106252       1 leaderelection.go:248] attempting to acquire leader lease vmware-system-tkg/tanzu-addons-manager-leader-lock...

# non-leader pod2
[ ~ ]# k logs tanzu-addons-controller-manager-556c8dddf6-984br -n vmware-system-tkg --tail 10
...
I0629 18:31:31.003785       1 leaderelection.go:248] attempting to acquire leader lease vmware-system-tkg/tanzu-addons-manager-leader-lock...
```

Create a new template antreaconfig in `test-ns-1`, verify only the leader controller reconciles the request:
```
[ ~ ]# k logs tanzu-addons-controller-manager-556c8dddf6-cq89s -n vmware-system-tkg --tail 10
...
I0629 18:35:21.642188       1 antreaconfig_controller.go:53] AntreaConfigController "msg"="Start reconciliation"
I0629 18:35:21.643107       1 antreaconfig_controller.go:82] AntreaConfigController "msg"="Cluster 'v1.23.5---vmware.1-tkg.1-zshippable'not found in 'test-ns-1'" "antreaconfig"={"Namespace":"test-ns-1","Name":"v1.23.5---vmware.1-tkg.1-zshippable"}

[ ~ ]# k logs tanzu-addons-controller-manager-556c8dddf6-72m82 -n vmware-system-tkg --tail 10
...
I0629 18:31:27.106252       1 leaderelection.go:248] attempting to acquire leader lease vmware-system-tkg/tanzu-addons-manager-leader-lock...
I0629 18:35:21.612268       1 logr.go:252] antreaconfig-resource "msg"="validate create"  "name"="v1.23.5---vmware.1-tkg.1-zshippable"

[ ~ ]# k logs tanzu-addons-controller-manager-556c8dddf6-984br -n vmware-system-tkg --tail 10
...
I0629 18:31:31.003785       1 leaderelection.go:248] attempting to acquire leader lease vmware-system-tkg/tanzu-addons-manager-leader-lock...
```

Delete current lead pod, verify new leader is elected:
```
[ ~ ]# k delete pod tanzu-addons-controller-manager-556c8dddf6-cq89s -n vmware-system-tkg --force
pod "tanzu-addons-controller-manager-556c8dddf6-cq89s" deleted

[ ~ ]# k get pod -n vmware-system-tkg | grep add
tanzu-addons-controller-manager-556c8dddf6-72m82         1/1     Running   0                6m52s
tanzu-addons-controller-manager-556c8dddf6-984br         1/1     Running   0                6m52s
tanzu-addons-controller-manager-556c8dddf6-hfwqw         1/1     Running   0                81s

# new leader pod
[ ~ ]# k logs tanzu-addons-controller-manager-556c8dddf6-984br -n vmware-system-tkg
I0629 18:31:31.003785       1 leaderelection.go:248] attempting to acquire leader lease vmware-system-tkg/tanzu-addons-manager-leader-lock...
...
I0629 18:36:51.113915       1 leaderelection.go:258] successfully acquired lease vmware-system-tkg/tanzu-addons-manager-leader-lock
...

# non-leader pods
[ ~ ]# k logs tanzu-addons-controller-manager-556c8dddf6-72m82 -n vmware-system-tkg --tail 10
...
I0629 18:31:27.106252       1 leaderelection.go:248] attempting to acquire leader lease vmware-system-tkg/tanzu-addons-manager-leader-lock...

[ ~ ]# k logs tanzu-addons-controller-manager-556c8dddf6-hfwqw -n vmware-system-tkg --tail 10
...
I0629 18:37:04.260751       1 leaderelection.go:248] attempting to acquire leader lease vmware-system-tkg/tanzu-addons-manager-leader-lock...
```

### Release note

```release-note
Enable leader election for Addons Controller
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: addons/main.go [hunk 1]
```diff
@@ -170,7 +182,10 @@ func main() {
 		Port:                   flags.webhookServerPort,
 		CertDir:                constants.WebhookCertDir,
 		LeaderElection:         flags.enableLeaderElection,
-		LeaderElectionID:       "5832a104.run.tanzu.addons",
+		LeaderElectionID:       flags.leaderElectionID,
+		LeaseDuration:          &flags.leaderElectionLeaseDuration,
+		RenewDeadline:          &flags.leaderElectionRenewDeadline,
+		RetryPeriod:            &flags.leaderElectionRetryPeriod,
```
Comment (maralavi) [inline addons/main.go:188]: Where do these parameters get used?

File: addons/main.go [hunk 2]
```diff
@@ -107,6 +111,14 @@ func parseAddonFlags(addonFlags *addonFlags) {
 	flag.BoolVar(&addonFlags.enableLeaderElection, "leader-elect", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	flag.StringVar(&addonFlags.leaderElectionID, "leader-elect-id", constants.AddonControllerManagerLeaderElectionResourceName,
+		"The name of the resource that leader election will use for holding the leader lock.")
+	flag.DurationVar(&addonFlags.leaderElectionLeaseDuration, "leader-elect-lease-duration", 15*time.Second,
+		"Interval at which non-leader candidates will wait to force acquire leadership")
+	flag.DurationVar(&addonFlags.leaderElectionRenewDeadline, "leader-elect-renew-deadline", 10*time.Second,
+		"Duration that the leading controller manager will retry refreshing leadership before giving up")
+	flag.DurationVar(&addonFlags.leaderElectionRetryPeriod, "leader-elect-retry-period", 2*time.Second,
```
Comment (maralavi) [inline addons/main.go:120]: Question: Which code uses these flags?

File: packages/addons-manager/bundle/config/values.yaml [hunk 3]
```diff
@@ -6,6 +6,7 @@ tanzuAddonsManager:
   namespace: tkg-system
   createNamespace: true
   deployment:
+    leaderElection: false
```
Comment (shyaamsn) [inline packages/addons-manager/bundle/config/values.yaml:9]: Currently addons-manager is installed using tanzu-framework and tkg composite package.

i.e. Installing tkg package would install framework package and installing tanzu-framework would install addons-manager and other packages.

So this `leaderElection: false` also has to be added in framework and tkg package

i.e. Please add leaderElection: false under

https://github.com/vmware-tanzu/tanzu-framework/blob/main/packages/tkg/bundle/config/values.yaml#L39
https://github.com/vmware-tanzu/tanzu-framework/blob/main/packages/framework/bundle/config/values.yaml#L28

File: addons/main.go (hunk 1 above)
Comment (HL-EverGreen) [reply addons/main.go:188]: Hi Marjan, thanks for the review. These parameters are in manager `Options` and will be used by `client-go` library, which will take care of the actual leader election logic.

File: addons/main.go (hunk 2 above)
Comment (HL-EverGreen) [reply addons/main.go:120]: These flags are used in the initialization of manager `Options`.

File: packages/addons-manager/bundle/config/values.yaml (hunk 3 above)
Comment (HL-EverGreen) [reply packages/addons-manager/bundle/config/values.yaml:9]: Good to know! Thanks for pointing that out. PR has been updated.

File: addons/main.go (hunk 1 above)
Comment (maralavi) [reply addons/main.go:188]: Thanks Leo for the clarification, I see these are defined here: https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/.

**Review by maralavi — APPROVED** [review]

**Comment by vijaykatam** [comment]
To take advantage of leader election, we need at least 2 replicas. Lets update the deployment package and test on a cluster to ensure that only 1 replica is processing requests.

**Comment by HL-EverGreen** [comment]
> To take advantage of leader election, we need at least 2 replicas. Lets update the deployment package and test on a cluster to ensure that only 1 replica is processing requests.

Hi @vijaykatam , the testing details can be seen at the PR description, I scaled the controller to 3 replicas and checked that there is only one replica processing requests.

**Review by vijaykatam — APPROVED** [review]

**Pull request merged**